### ISSUE TITLE
Running PredicateBuilder Queries when nodes join is causing timeouts and cluster crash.

### ISSUE THREAD

**Issue by kraythen** [hazelcast/hazelcast#6242]
We have many queries that use PredicateBuilder queries such as the one below. These predicates run a lot during the normal usage of our application. However, when we have a new node join the cluster while the predicates are joining we get an info log warning and then the whole cluster stops responding and sometimes fails completely. This happens much more when we used submitToKeyOwner() and are running the predicate inside that task. The log and sample code are below.

```
Sep 17 12:30:52 staging_i-47ba8f82 application-9001.log:  2015-09-17 17:30:52 +0000 - [INFO] - [10.15.1.136]:5701 [staging] [3.5.2] Partition assignments changed while executing query: PredicateBuilder{
Sep 17 12:30:52 staging_i-47ba8f82 application-9001.log:  (userId=587 AND itemId IN (222375,222374,222373,222379,222378,222377,222376,222383,222382,222381,222380,222387,222386,222385,222384,222389,222388))
```

``` java
public static Set<Integer> findCachedOrderIdsByUserItem(Long userId,  Collection<Integer> itemIds) {
  final EntryObject e = new PredicateBuilder().getEntryObject();
  return hazelcast.getMap("orders").keySet(e.get("userId").equal(userId)
      .and(e.get("itemId").in(itemIds.toArray(new Integer[itemtIds.size()]))));
}
```

We are running version 3.5.2 on AWS.

**Comment by vbekiaris** [comment]
@kraythen thanks for reporting this issue, does it still stand for you? Would you be able to test with latest released version (3.6.4) or 3.7-EA? There have been several changes in handling of migrations and querying since 3.5.2.

**Comment by mdogan** [comment]
Queries infrastructure changed significantly since 3.5. Please test with the latest version and re-open issue if you encounter the same problem again.
